File: ios/HandReceipt/ViewModels/TransfersViewModel.swift
import Foundation
import Combine
import SwiftUI // Needed for UUID

@MainActor
class TransfersViewModel: ObservableObject {
    
    // MARK: - State Enums
    enum LoadingState: Equatable { // Add Equatable conformance
        case idle
        case loading
        case success([Transfer])
        case error(String)

        // Manually implement Equatable for enums with associated values
        static func == (lhs: LoadingState, rhs: LoadingState) -> Bool {
            switch (lhs, rhs) {
            case (.idle, .idle): return true
            case (.loading, .loading): return true
            // For success, we might compare based on content if needed, but for simple state checking,
            // just checking if both are .success might be enough. Comparing arrays can be complex.
            case (.success(let lTransfers), .success(let rTransfers)): return lTransfers == rTransfers // Compare transfer arrays
            case (.error(let lMsg), .error(let rMsg)): return lMsg == rMsg
            default: return false // Different cases are not equal
            }
        }
    }
    
    enum ActionState: Equatable { // Conform to Equatable for easier state comparison/animation
        case idle
        case loading
        case success(String) // Success message (e.g., "Transfer Approved")
        case error(String) // Error message
    }
    
    enum FilterDirection: String, CaseIterable, Identifiable {
        case all = "All"
        case incoming = "Incoming"
        case outgoing = "Outgoing"
        var id: String { self.rawValue }
    }
    
    enum FilterStatus: String, CaseIterable, Identifiable {
        case pending = "Pending"
        case all = "All"
        case approved = "Approved"
        case rejected = "Rejected"
        case cancelled = "Cancelled"
        var id: String { self.rawValue }
    }

    // MARK: - Published Properties
    @Published var loadingState: LoadingState = .idle
    @Published var actionState: ActionState = .idle // For approve/reject actions
    @Published var selectedDirectionFilter: FilterDirection = .all
    @Published var selectedStatusFilter: FilterStatus = .pending
    
    // MARK: - Dependencies
    private let apiService: APIServiceProtocol
    private var cancellables = Set<AnyCancellable>()
    
    // TODO: We need access to the current user's ID to implement direction filtering correctly.
    // This might come from an injected AuthViewModel or a shared session manager.
    // For now, the computed property will ignore the direction filter.
    var currentUserId: Int? = nil // Changed from UUID?
    
    // Computed property for filtered transfers
    var filteredTransfers: [Transfer] {
        guard case .success(let transfers) = loadingState else { return [] }
        
        return transfers.filter { transfer in
            // Status Filter
            let statusMatch = (selectedStatusFilter == .all || transfer.status.rawValue.lowercased() == selectedStatusFilter.rawValue.lowercased())
            
            // Direction Filter (requires currentUserId)
            var directionMatch = true // Default to true if no filter or ID unavailable
            if let userId = currentUserId, selectedDirectionFilter != .all {
                if selectedDirectionFilter == .incoming {
                    directionMatch = (transfer.toUserId == userId)
                } else { // .outgoing
                    directionMatch = (transfer.fromUserId == userId)
                }
            }
            
            return statusMatch && directionMatch
        }
    }

    // MARK: - Initialization
    init(apiService: APIServiceProtocol = APIService(), currentUserId: Int? = nil) { // Changed from UUID?
        self.apiService = apiService
        self.currentUserId = currentUserId // Set the current user ID
        print("TransfersViewModel initialized. Current User ID: \(currentUserId?.description ?? "N/A")") // Use .description for Int?
        
        // Reload transfers when filters change
        Publishers.CombineLatest($selectedDirectionFilter, $selectedStatusFilter)
            .debounce(for: .milliseconds(100), scheduler: RunLoop.main) // Small debounce
            .sink { [weak self] _, _ in
                self?.fetchTransfers()
            }
            .store(in: &cancellables)
        
        // Initial Fetch
        fetchTransfers()
    }

    // MARK: - API Calls
    func fetchTransfers() {
        print("TransfersViewModel: Fetching transfers (Status: \(selectedStatusFilter.rawValue), Direction: \(selectedDirectionFilter.rawValue))")
        // Always fetch ALL for now, filtering happens in computed property
        // TODO: Update API call if backend supports server-side filtering for status/direction
        let statusParam = selectedStatusFilter == .all ? nil : selectedStatusFilter.rawValue.uppercased()
        let directionParam: String? = nil // Fetch all directions, filter locally based on currentUserId

        loadingState = .loading
        actionState = .idle // Clear action state on refresh
        
        Task {
            do {
                let transfers = try await apiService.fetchTransfers(status: statusParam, direction: directionParam)
                print("TransfersViewModel: Successfully fetched \(transfers.count) transfers from API.")
                loadingState = .success(transfers)
            } catch {
                print("TransfersViewModel: Error fetching transfers - \(error.localizedDescription)")
                loadingState = .error("Failed to load transfers: \(error.localizedDescription)")
            }
        }
    }
    
    func approveTransfer(transferId: Int) { // Changed from UUID
        print("TransfersViewModel: Approving transfer ID \(transferId)")
        actionState = .loading
        clearActionTimer?.cancel() // Cancel any pending clear timer
        
        Task {
            do {
                _ = try await apiService.approveTransfer(transferId: transferId) // Pass Int
                print("TransfersViewModel: Successfully approved transfer ID \(transferId)")
                actionState = .success("Transfer Approved")
                // Refresh the list to show updated status
                fetchTransfers()
                // Schedule state clear after delay
                scheduleActionStateClear()
            } catch {
                print("TransfersViewModel: Error approving transfer ID \(transferId) - \(error.localizedDescription)")
                actionState = .error("Approval failed: \(error.localizedDescription)")
                scheduleActionStateClear()
            }
        }
    }
    
    func rejectTransfer(transferId: Int) { // Changed from UUID
        print("TransfersViewModel: Rejecting transfer ID \(transferId)")
        actionState = .loading
        clearActionTimer?.cancel()
        
        Task {
            do {
                _ = try await apiService.rejectTransfer(transferId: transferId) // Pass Int
                print("TransfersViewModel: Successfully rejected transfer ID \(transferId)")
                actionState = .success("Transfer Rejected")
                // Refresh the list to show updated status
                fetchTransfers()
                scheduleActionStateClear()
            } catch {
                print("TransfersViewModel: Error rejecting transfer ID \(transferId) - \(error.localizedDescription)")
                actionState = .error("Rejection failed: \(error.localizedDescription)")
                scheduleActionStateClear()
            }
        }
    }
    
    // MARK: - Helpers
    private var clearActionTimer: AnyCancellable?
    
    private func scheduleActionStateClear(delay: TimeInterval = 3.0) {
        clearActionTimer?.cancel()
        clearActionTimer = Just(())
            .delay(for: .seconds(delay), scheduler: RunLoop.main)
            .sink { [weak self] _ in
                // Only clear if it's still in a success/error state
                if self?.actionState != .loading && self?.actionState != .idle {
                    print("Clearing action state.")
                    self?.actionState = .idle
                }
            }
    }
    
    deinit {
        clearActionTimer?.cancel()
        cancellables.forEach { $0.cancel() }
        print("TransfersViewModel deinitialized")
    }
} 

File: android/app/src/main/java/com/example/handreceipt/viewmodels/TransfersViewModel.kt
package com.example.handreceipt.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.handreceipt.data.model.Transfer
import com.example.handreceipt.data.model.TransferStatus
import com.example.handreceipt.data.model.UserSummary // Assuming UserSummary exists for filtering
import com.example.handreceipt.data.network.ApiService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.launch
import retrofit2.HttpException
import java.io.IOException
import java.util.UUID
import javax.inject.Inject

// --- State Enums ---
sealed class TransfersLoadingState {
    object Idle : TransfersLoadingState()
    object Loading : TransfersLoadingState()
    data class Success(val allTransfers: List<Transfer>) : TransfersLoadingState()
    data class Error(val message: String) : TransfersLoadingState()
}

sealed class TransferActionState {
    object Idle : TransferActionState()
    object Loading : TransferActionState()
    data class Success(val message: String) : TransferActionState()
    data class Error(val message: String) : TransferActionState()
}

// --- Filter Enums ---
enum class TransferDirectionFilter(val displayName: String) {
    ALL("All"),
    INCOMING("Incoming"),
    OUTGOING("Outgoing")
}

enum class TransferStatusFilter(val displayName: String, val apiValue: String?) {
    PENDING("Pending", "PENDING"),
    ALL("All", null),
    APPROVED("Approved", "APPROVED"),
    REJECTED("Rejected", "REJECTED"),
    CANCELLED("Cancelled", "CANCELLED")
}

@HiltViewModel
class TransfersViewModel @Inject constructor(
    private val apiService: ApiService,
    // TODO: Inject a way to get the current user ID (e.g., from a SessionManager or AuthRepository)
    private val currentUserId: UUID? // Placeholder - Needs proper injection
) : ViewModel() {

    // --- State Flows ---
    private val _loadingState = MutableStateFlow<TransfersLoadingState>(TransfersLoadingState.Idle)
    val loadingState: StateFlow<TransfersLoadingState> = _loadingState.asStateFlow()

    private val _actionState = MutableStateFlow<TransferActionState>(TransferActionState.Idle)
    val actionState: StateFlow<TransferActionState> = _actionState.asStateFlow()

    private val _selectedDirectionFilter = MutableStateFlow(TransferDirectionFilter.ALL)
    val selectedDirectionFilter: StateFlow<TransferDirectionFilter> = _selectedDirectionFilter.asStateFlow()

    private val _selectedStatusFilter = MutableStateFlow(TransferStatusFilter.PENDING)
    val selectedStatusFilter: StateFlow<TransferStatusFilter> = _selectedStatusFilter.asStateFlow()
    
    private val _filteredTransfers = MutableStateFlow<List<Transfer>>(emptyList())
    val filteredTransfers: StateFlow<List<Transfer>> = _filteredTransfers.asStateFlow()
    
    private var clearActionJob: Job? = null

    init {
        // Combine loading state and filters to update filteredTransfers
        combine(
            loadingState,
            selectedDirectionFilter,
            selectedStatusFilter
        ) { loadState, direction, status ->
            if (loadState is TransfersLoadingState.Success) {
                filterTransfersLocally(loadState.allTransfers, direction, status)
            } else {
                emptyList()
            }
        }.debounce(50) // Small debounce for filter changes
         .launchIn(viewModelScope)
         .also { job -> 
            job.invokeOnCompletion { _filteredTransfers.value = emptyList() } // Clear on scope cancellation
        } // Assign to _filteredTransfers indirectly via collect
        .let { flow -> 
             viewModelScope.launch { flow.collect { _filteredTransfers.value = it } }
        }

        // Fetch transfers when filters change
        combine(selectedDirectionFilter, selectedStatusFilter) { _, _ -> Unit }
            .debounce(100) // Debounce filter changes before fetching
            .launchIn(viewModelScope)
            .let { flow ->
                viewModelScope.launch { flow.collect { fetchTransfers() } }
            }
        
        // Initial fetch
        fetchTransfers()
    }

    // --- Filter Logic ---
    private fun filterTransfersLocally(
        allTransfers: List<Transfer>,
        direction: TransferDirectionFilter,
        status: TransferStatusFilter
    ): List<Transfer> {
        println("Filtering locally: ${allTransfers.size} items, Dir: $direction, Status: $status, User: ${currentUserId?.toString() ?: "N/A"}")
        return allTransfers.filter { transfer ->
            val statusMatch = (status == TransferStatusFilter.ALL || transfer.status.name == status.apiValue)
            val directionMatch = when (direction) {
                TransferDirectionFilter.ALL -> true
                TransferDirectionFilter.INCOMING -> transfer.toUserId == currentUserId
                TransferDirectionFilter.OUTGOING -> transfer.fromUserId == currentUserId
            }
            statusMatch && directionMatch
        }
    }
    
    fun setDirectionFilter(filter: TransferDirectionFilter) {
        _selectedDirectionFilter.value = filter
    }

    fun setStatusFilter(filter: TransferStatusFilter) {
        _selectedStatusFilter.value = filter
    }

    // --- API Calls ---
    fun fetchTransfers() {
        _loadingState.value = TransfersLoadingState.Loading
        _actionState.value = TransferActionState.Idle // Clear action state
        viewModelScope.launch {
            try {
                println("Fetching transfers from API. Status: ${selectedStatusFilter.apiValue}, Direction: ALL (local filtering)")
                // Fetch ALL directions, filter locally based on currentUserId
                // Fetch based on status filter from API
                val response = apiService.getTransfers(status = selectedStatusFilter.apiValue, direction = null)
                if (response.isSuccessful && response.body() != null) {
                    _loadingState.value = TransfersLoadingState.Success(response.body()!!)
                    println("Fetched ${response.body()!!.size} transfers successfully.")
                } else {
                    val errorMsg = "Error ${response.code()}: ${response.errorBody()?.string() ?: "Failed to load"}"
                    _loadingState.value = TransfersLoadingState.Error(errorMsg)
                    println("Error fetching transfers: $errorMsg")
                }
            } catch (e: HttpException) {
                 val errorMsg = "Network error: ${e.message()}"
                _loadingState.value = TransfersLoadingState.Error(errorMsg)
                 println("Error fetching transfers: $errorMsg")
            } catch (e: IOException) {
                 val errorMsg = "Connection error. Please check network."
                _loadingState.value = TransfersLoadingState.Error(errorMsg)
                 println("Error fetching transfers: $errorMsg")
            } catch (e: Exception) {
                 val errorMsg = "An unexpected error occurred: ${e.localizedMessage}"
                _loadingState.value = TransfersLoadingState.Error(errorMsg)
                 println("Error fetching transfers: $errorMsg")
            }
        }
    }

    fun approveTransfer(transferId: String) {
        performAction("Approving", transferId) { apiService.approveTransfer(it) }
    }

    fun rejectTransfer(transferId: String) {
        performAction("Rejecting", transferId) { apiService.rejectTransfer(it) }
    }

    private fun performAction(actionName: String, transferId: String, apiCall: suspend (String) -> Response<Transfer>) {
        _actionState.value = TransferActionState.Loading
        clearActionJob?.cancel()
        println("$actionName transfer ID: $transferId")
        viewModelScope.launch {
            try {
                val response = apiCall(transferId)
                if (response.isSuccessful) {
                    _actionState.value = TransferActionState.Success("Transfer ${actionName.dropLast(3)}ed")
                    println("Successfully ${actionName.dropLast(3)}ed transfer ID $transferId")
                    fetchTransfers() // Refresh list
                    scheduleActionStateClear()
                } else {
                    val errorMsg = "Error ${response.code()}: ${response.errorBody()?.string() ?: "Action failed"}"
                    _actionState.value = TransferActionState.Error(errorMsg)
                    println("Error ${actionName.dropLast(1)} transfer $transferId: $errorMsg")
                    scheduleActionStateClear()
                }
            } catch (e: HttpException) {
                val errorMsg = "Network error: ${e.message()}"
                _actionState.value = TransferActionState.Error(errorMsg)
                 println("Error ${actionName.dropLast(1)} transfer $transferId: $errorMsg")
                scheduleActionStateClear()
            } catch (e: IOException) {
                 val errorMsg = "Connection error. Please check network."
                _actionState.value = TransferActionState.Error(errorMsg)
                 println("Error ${actionName.dropLast(1)} transfer $transferId: $errorMsg")
                scheduleActionStateClear()
            } catch (e: Exception) {
                 val errorMsg = "An unexpected error occurred: ${e.localizedMessage}"
                _actionState.value = TransferActionState.Error(errorMsg)
                 println("Error ${actionName.dropLast(1)} transfer $transferId: $errorMsg")
                scheduleActionStateClear()
            }
        }
    }

    private fun scheduleActionStateClear(delayMillis: Long = 3000) {
        clearActionJob?.cancel()
        clearActionJob = viewModelScope.launch {
            delay(delayMillis)
            if (_actionState.value !is TransferActionState.Idle && _actionState.value !is TransferActionState.Loading) {
                _actionState.value = TransferActionState.Idle
            }
        }
    }
} 